File: leetcode/algorithms/easy/merge-two-sorted-lists.js
/*
    leetcode
    category: "Algorithms"
    problem title: "Merge Two Sorted Lists"
*/

/**
 * Definition for singly-linked list.
 * function ListNode(val) {
 *     this.val = val;
 *     this.next = null;
 * }
 */
/**
 * @param {ListNode} l1
 * @param {ListNode} l2
 * @return {ListNode}
 */
var mergeTwoLists = function(l1, l2) {
    let listHead = null,tail=null;
    while(l1!=null || l2!=null){
        let node = new ListNode(null);
        if(l1 && (!l2 || (l1.val<=l2.val))){
            node.val = l1.val;
            l1 = l1.next;
        } else {
            node.val = l2.val;
            l2 = l2.next;
        }
        
        if(!listHead){
            listHead=node;
            tail=node;
        } else {
            tail.next=node;
            tail=node;
        }
        
    }
    
    return listHead;
};

File: leetcode/algorithms/linked-list/merge-two-sorted-lists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */
class Solution {
    public ListNode mergeTwoLists(ListNode list1, ListNode list2) {
        if(list1==null){
            return list2;
        }
        if(list2==null){
            return list1;
        }
        ListNode head = null, tail = null;
        
        while(list1!=null && list2!=null){
            ListNode nextNode = list1.val<=list2.val ? list1 : list2;
            if(head==null){
                head = nextNode;
                tail = nextNode;
            } else {
                tail.next = nextNode;
                tail = nextNode;
            }
            if(nextNode==list1){
                list1 = list1.next;
            } else {
                list2 = list2.next;
            }
        }
        
        ListNode nonNullNode = list1==null ? list2 : list1;
        
        tail.next = nonNullNode;
        return head;
        
    }
}